[lldb] Add support for updating string during debug process

kpdev opened the pull request:
Before this update strings has only DataFormatters, and therefore the way they were printed was slightly different (see comment here: https://github.com/llvm/llvm-project/blob/main/lldb/source/DataFormatters/FormatManager.cpp#L498), so tests which rely on this formatting also changed. std::string/std::wstring/std::u16(32)string synthetic frontend implemented Also tests for the frontend added.

~~

Huawei RRI, OS Lab

kpdev commented:
Depends on: https://github.com/llvm/llvm-project/pull/67309#issuecomment-1740657973

jimingham commented on lldb/test/API/functionalities/data-formatter/data-formatter-stl/generic/multimap/TestDataFormatterGenericMultiMap.py:
```diff
@@ -231,23 +244,39 @@ def cleanup():
-                '[1] = (first = 2, second = "smart")',
-                '[2] = (first = 3, second = "!!!")',
-                '[3] = (first = 85, second = "goofy")',
+                'size=4',
```
What is it about this change that is defeating the ValueObject printer from compressing this output onto one line?  It looks like the contents that get printed are the same, so there's something about switching from a Summary provider to a child provider that's causing problems.  We should fix that as people are really picky about variable printing being as space efficient as possible.

jimingham commented:
BTW, I have no problem with the general direction of this change.  It makes a lot more sense to ask a synthetic child provider to change a value - since it does represent the value of the ValueObject - rather than the summary which is just some free-form text.  And being able to change characters in a string seems a reasonable thing to do, so switching the std::string comprehension from a Summary provider to a Synthetic Child Provider is the way to do that.  So that part if fine.  

But std::strings abound in code, and so if we're going to make this change I don't think we can make that printing less space efficient, which this change seems to have done.  We should figure out why that's the case and fix for this to be a really good change.

kpdev commented:
> What is it about this change that is defeating the ValueObject printer from compressing this output onto one line? It looks like the contents that get printed are the same, so there's something about switching from a Summary provider to a child provider that's causing problems. We should fix that as people are really picky about variable printing being as space efficient as possible.

> To clear up terminology... Strings had data formatters before AND after this change. The difference is that you've switched from a "Summary Provider" data formatter to a "Synthetic Child Provider" data formatter.
> 
> It looks like you've made the printing of std::strings less space efficient. That shouldn't be necessary, and isn't desirable. We should figure out why that's happening and fix that before this change is going to not cause complaints.

As there is mentioned in a comment for `FormatManager::ShouldPrintAsOneLiner` ( https://github.com/llvm/llvm-project/blob/main/lldb/source/DataFormatters/FormatManager.cpp#L498 ):

```
    // if we decided to define synthetic children for a type, we probably care
    // enough to show them, but avoid nesting children in children
```

So, there is a condition for that:

```
      // but if we only have them to provide a value, keep going
      if (!synth_sp->MightHaveChildren() &&
          synth_sp->DoesProvideSyntheticValue())
        is_synth_val = true;
      else
        return false;
```

This patch adds StringSynthetic and this synthetic `MightHaveChildren()` is `true`, therefore `ShouldPrintAsOneLiner` returns `false`. And the printer will use `"()"` or `"{\n ... \n}"` according to whether we return `true` or `false` from this function. If we would like to avoid unnecessary output we probably may ask SyntheticFrontend directly if it would like to print expanded info or not, what do you think about it?

kpdev commented:
> BTW, I have no problem with the general direction of this change. It makes a lot more sense to ask a synthetic child provider to change a value - since it does represent the value of the ValueObject - rather than the summary which is just some free-form text. And being able to change characters in a string seems a reasonable thing to do, so switching the std::string comprehension from a Summary provider to a Synthetic Child Provider is the way to do that. So that part if fine.
> 
> But std::strings abound in code, and so if we're going to make this change I don't think we can make that printing less space efficient, which this change seems to have done. We should figure out why that's the case and fix for this to be a really good change.

So, is it Ok to use current `SetValueFromCString` api from ValueObject to ask synthetic provider to update the underlying string?
You mentioned previously that we may add `SetSummaryFromCString` api - in fact currently this is what I am doing - changing the whole string through its summary (please check attached gif for example).
But the problem with the new API is the same as for the changing characters through `SBValue::GetData` - IDE doesn't support it

jimingham commented:
> On Oct 5, 2023, at 1:05 AM, Pavel Kosov ***@***.***> wrote:
> 
> 
> BTW, I have no problem with the general direction of this change. It makes a lot more sense to ask a synthetic child provider to change a value - since it does represent the value of the ValueObject - rather than the summary which is just some free-form text. And being able to change characters in a string seems a reasonable thing to do, so switching the std::string comprehension from a Summary provider to a Synthetic Child Provider is the way to do that. So that part if fine.
> 
> But std::strings abound in code, and so if we're going to make this change I don't think we can make that printing less space efficient, which this change seems to have done. We should figure out why that's the case and fix for this to be a really good change.
> 
> So, is it Ok to use current SetValueFromCString api from ValueObject to ask synthetic provider to update the underlying string?
> 
I'm going to use SCP for "synthetic child provider" as I was getting tired of typing it out...  And I think you know most of what I'm saying here, but I wanted to make sure we are talking about it the same way...

Since "SetValueFromCString" is an API on ValueObject, we are free to give that whatever meaning makes sense.

However, I think it's confusing here to start with the model that ValueObjects with SCPs have "underlying strings".  The vast majority of data types represented by synthetic child providers have no single underlying string, they are mapping aggregates to aggregates or reinterpreting various regions (contiguous or not) of memory as though it were an aggregate of some sort.  So there will for the most part be no "underlying string".

More generally, we don't really do "SetValueFromCString" for aggregate types, only for scalar values.  So I don't see how SetValueFromCString is a useful general API on the aggregate type that actually holds the SCP.  You really have to specify the particular child you want to change.

What does makes pretty clear sense is that if SetValueFromCString is called on one of the synthetic children produced by the ValueObject holding the SCP, the child ValueObject should ask the parent's SCP: "can you change my value to the incoming string".  The SCP API would have to grow another interface "SetValueOfChildFromCString" to support this.  The SCP has to know how to do that in a way that round-trips (i.e. if I make a second ValueObject from the same variable, it will also show the changed value).  If it can do that, then it will make the change.  Note, it may very well have to change others of the synthetic child values to make that happen, depending on how the child is generated from the underlying data being represented.  It's also not guaranteed to be 1-1, since there might be many ways to change the underlying data that produce the requested change on the synthetic child.  We should probably decide whether "I can change to represent this, but not uniquely" should be an error or not.

> You mentioned previously that we may add SetSummaryFromCString api - in fact currently this is what I am doing - changing the whole string through its summary (please check attached gif for example).
> 
I think this is also a useful API though it is a bit confusingly named.  Sadly I can't think of a better one that isn't overly verbose.  You aren't really setting the summary - that's not a terribly useful operation if the underlying value doesn't produce that new summary when asked for it again.  It's really "ask the summary provider if it knows how to change the underlying data such that it will produce this new summary string".  But anyway, we can make this clear in the API docs.

By the way, as a side note, watching the part of your example where you change the raw string guts, it looks like we don't update summaries of a ValueObject if one of its children gets changed.  Be good to file bug on that one so we don't forget.

> But the problem with the new API is the same as for the changing characters through SBValue::GetData - IDE doesn't support it
> 
In the case of std::strings, which seems to be your primary motivator here, I think considering the string value to be the summary makes the most sense.  There really isn't a scalar type of "a buffer of memory starting at 0x123 and ending sometime later (maybe signaled by there being a `\0` somewhere)."  So representing std::string with a value providing SCP is not at all natural.  Also in this case, it's useful to be able to see its raw contents easily in the UI as you were doing in your example, and if you swapped the std::string ValueObject for some kind of a "C-String" ValueObject, it would be awkward to get back to the raw fields.

The UI knows that it is inserting a string it got from GetSummary in that slot in the UI, so it can easily know to call SetSummaryFromCString when that slot is edited.  I am not in favor of using a strained model for as ubiquitous a data type as std::string in lldb just because we don't want to have to teach the UI how to handle it properly.

Jim
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/67782#issuecomment-1748331579>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW7EA2Y2567SD5JCJMTX5ZS6PAVCNFSM6AAAAAA5MH6OUGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONBYGMZTCNJXHE>.
> You are receiving this because you commented.
>

kpdev committed:
[LLDB] Add ability to update string during debugging

This is the last patch needed for adding an ability to update std::string/wstring/etc during debug process.

std::string/std::wstring/std::u16(32)string synthetic frontend implemented
Also tests for the frontend added.

~~

Huawei RRI, OS Lab

kpdev commented:
Hello, @jimingham , first of all - sorry for the long delay in reply.

I carefully read through all your messages once again and found, that I totally misunderstood some places, i.e. "Why isn't it enough to have the ValueObjectSynthetic's SetValueFromCString do this" - I think it should be enough, I'll fix it.

> Secondly, lldb only supports changing scalars because it's hard to give meaning to "changing an aggregate object".  You have to change it in situ - which really means changing the values of its contents, i.e. its children - or you will end up invalidating code that relies on the location of the aggregate

I agree with this, but I think there is not so much aggregates for which it makes sense to change their length as for strings (at least in libcxx) - I mean it is natural to update whole string to completely different value, but it is not natural to do so for e.g. vector. In case of strings, one might want to set string longer, than the one he has now for the debug purposes, so this will indeed invalidate code, that relies on the pointers that was obtained through `.data` or `.c_str` methods, but it is the programmer responsibility to care about this. This it the same behaviour as if you change your string in the program - you should update your pointers.

> However, I think it's confusing here to start with the model that ValueObjects with SCPs have "underlying strings".  

Sorry, I think that I didn't express my thoughts carefully, by the underlying string I didn't mean, that we have some string in the SCP or ValueObjects, I meant the strings in the code that is under debug. 

> By the way, as a side note, watching the part of your example where you change the raw string guts, it looks like we don't update summaries of a ValueObject if one of its children gets changed.  Be good to file bug on that one so we don't forget.

I'm not sure that this bug might be reproduced without the string example, I don't know which type have the summary which represent all its children. Is it ok, to file a bug with current strings example or how to do it better?

> In the case of std::strings, which seems to be your primary motivator here, I think considering the string value to be the summary makes the most sense  

And in the end, may I kindly ask you to clarify your position about these changes please? Do you suggest to return to the `SetSummaryFromCString` API?

kpdev commented:
@jimingham Please take a look: https://github.com/llvm/llvm-project/pull/67782#issuecomment-2047369473

jimingham commented on lldb/source/Plugins/Language/CPlusPlus/LibCxxString.cpp:
```diff
@@ -0,0 +1,171 @@
+    Target *target = exe_ctx.GetTargetPtr();
+    StackFrame *frame = exe_ctx.GetFramePtr();
+
+    if (target && frame) {
```
I don't think this is the right way to implement this.  By relying on the base name of the backend and using that directly you will only be able to handle cases where the std::string is a local variable or function argument.  It won't work on std::strings inside other objects - you would need the full path.  It also won't work for globals, which don't have frames, even though there's no reason it shouldn't.

A better approach would be to get the address of the object you are trying to change, and using that address properly cast to do the assignment.  That way you wouldn't have to care how that value got referred to.  

For the most part we try not to run expressions in the ValueObject code, but setting values is less likely to happen for large objects, or to lots of variables at once the way rendering their values/Summaries, etc does.  But you certainly shouldn't run expressions on a ValueObject's behalf that allow more than one thread to run.  Otherwise you get into the state where I stop in thread A, see something interesting, go to thread B and change a value and switch back to thread A and it's not where I left it!

jimingham commented:
> Hello, @jimingham , first of all - sorry for the long delay in reply.
> 
> I carefully read through all your messages once again and found, that I totally misunderstood some places, i.e. "Why isn't it enough to have the ValueObjectSynthetic's SetValueFromCString do this" - I think it should be enough, I'll fix it.
> 
> > Secondly, lldb only supports changing scalars because it's hard to give meaning to "changing an aggregate object".  You have to change it in situ - which really means changing the values of its contents, i.e. its children - or you will end up invalidating code that relies on the location of the aggregate
> 
> I agree with this, but I think there is not so much aggregates for which it makes sense to change their length as for strings (at least in libcxx) - I mean it is natural to update whole string to completely different value, but it is not natural to do so for e.g. vector. In case of strings, one might want to set string longer, than the one he has now for the debug purposes, so this will indeed invalidate code, that relies on the pointers that was obtained through `.data` or `.c_str` methods, but it is the programmer responsibility to care about this. This it the same behaviour as if you change your string in the program - you should update your pointers.

It's not just pointers, of course.  If you've done:

size_t string_len = my_string.size();

then string_len will be wrong.  But I agree, that's more the lookout of whoever is messing with their strings while running.

> 
> > However, I think it's confusing here to start with the model that ValueObjects with SCPs have "underlying strings".
> 
> Sorry, I think that I didn't express my thoughts carefully, by the underlying string I didn't mean, that we have some string in the SCP or ValueObjects, I meant the strings in the code that is under debug.
> 
> > By the way, as a side note, watching the part of your example where you change the raw string guts, it looks like we don't update summaries of a ValueObject if one of its children gets changed.  Be good to file bug on that one so we don't forget.
> 
> I'm not sure that this bug might be reproduced without the string example, I don't know which type have the summary which represent all its children. Is it ok, to file a bug with current strings example or how to do it better?
> 
> > In the case of std::strings, which seems to be your primary motivator here, I think considering the string value to be the summary makes the most sense
> 
> And in the end, may I kindly ask you to clarify your position about these changes please? Do you suggest to return to the `SetSummaryFromCString` API?

The more I think about it, the more it seems that SetSummaryFromCString is just the wrong API to use to change the value of something.  After all, Summaries don't actually represent the "value" of the ValueObject, they are just whatever bits of that value the Summary Provider decided to extract, plus whatever additional text they wanted to provide.  Summaries may not even be about the value at all, for instance for checked pointers they might just say the checked state.  It's clear that's not the entity you would ask to change that object's value.

`std::string` and `char *` seem different because we show the string contents in the summary, and people like to pretend that the "value" of a std::string or char * is the extracted string contents, though that's not really true.

So I don't think trying to change underlying values makes sense as part of the Summary feature.

Changing values of a ValueObject with a Synthetic Child Provider definitely makes sense if directed to the children the synthetic child provider shows.  If the SCP for some value produces an `int` child named `my_made_up_child`, then it would make sense to ask it if it can change the value of whatever underlies `my_made_up_child`.  That also has a straight-forward interface, you get a SCP's child, and set its value, which the SCP Frontend intercepts, and does what is needed to change the value if it can.

If we were presenting the std::string contents as an array of char's then changing the individual characters by addressing the members of the array would make sense.  Note, that's also the only interface that will work for std::string's in full generality because std::strings aren't required to hold null-terminated C-strings.  

I'm a little bothered by the notion of changing the whole of a SyntheticValue by providing some C-string.  That makes sense for a restricted representation of the new values of something that stores string-y objects, but doesn't make sense without further syntax for anything else.  It's also not the right implementation internally for changing the child provided by an SCP, since it that case you'd call SetValueFromCString on the child, and route that to the parent.  

But having this in place wouldn't preclude adding machinery to get the SCP to effect a change for one of the children they produced.  And I guess we can make some attempt to define what this actually means the second time someone uses it.
